pytest_archon/rule.py

```diff
@@ -147,68 +147,100 @@ def check(
-
-            for constraint in self.required:
-                matches = {imp for imp in imports if fnmatch(imp, constraint)}
-                check.is_true(
-                    matches,
+            for constraint in self._check_required_constraints(candidate, import_map):
+                log_failure(
                     _fmt_rule(
                         rule_name,
                         rule_comment,
-                        f"module '{c}' is missing REQUIRED imports matching pattern /{constraint}/",
+                        f"module '{candidate}' is missing REQUIRED imports matching pattern /{constraint}/",
                     ),
                 )
-            for constraint in self.forbidden:
-                matches = {imp for imp in imports if fnmatch(imp, constraint)}
-                check.is_false(
-                    matches,
+
+            for match, constraint, seen in self._check_forbidden_constraints(candidate, import_map):
+                log_failure(
                     _fmt_rule(
                         rule_name,
                         rule_comment,
-                        f"module '{c}' has FORBIDDEN imports:\n{matches} (matched by /{constraint}/)",
+                        f"module '{candidate}' has FORBIDDEN imports:\n{match} (matched by /{constraint}/), "
+                        f"through modules {' ↣ '.join(seen)}.",
                     ),
                 )
 
+    def _check_required_constraints(self, module: str, all_imports: ImportMap):
+        for constraint in self.required:
+            if not any(imp for imp in recurse_imports(module, all_imports) if fnmatch(imp, constraint)):
+                yield constraint
+
+    def _check_forbidden_constraints(
+        self,
+        module: str,
+        all_imports: ImportMap,
+    ):
+        seen = set()
+
+        def check(mod, path):
+            if mod in seen or mod not in all_imports:
+                return
+
+            imports = all_imports[mod]
+            new_path = path + [mod]
+            seen.add(mod)
+
+            for constraint in self.forbidden:
+                for imp in imports:
+                    if any(fnmatch(imp, ignore) for ignore in self.ignored):
+                        continue
+                    if fnmatch(imp, constraint):
+                        yield (imp, constraint, new_path)
+                    else:
+                        yield from check(imp, new_path)
+
+        yield from check(module, [])
+
 
 def _fmt_rule(name, comment, text):
     res = f"RULE {name}: {text}"
     if comment:
         res += f"\n({comment})"
     return res
+
+
+def recurse_imports(module: str, all_imports: ImportMap):
+    seen = set()
+
+    def recurse(mod):
+        if mod in seen or mod not in all_imports:
+            return
+
+        imports = all_imports[mod]
+        seen.add(mod)
+        for imp in imports:
+            yield from imports
```

Comment:
are you sure that the `yield from imports` should be inside the for loop?

Reply:
This looks a bit redundant, indeed.